Deluxe Edition

UOBW opened the pull request:
* Added a setting to enable Deluxe Edition
* The wildfire no longer blanks outsiders if the setting is enabled
* The phoenix logic uses the logic of the deluxe edition if the setting is enabled ("Phoenix is immune to the Book of Changes and may not Blank or be Blanked by any other card" instead of "If this card is BLANKED for any reason, its base strength is reduced to 0 but it retains its suits")
* Reordered the settings page to better fit in the new setting and added some horizontal dividers to improve organization
* Updated the disclaimer on the settings page

I tested the changes on my phone and everything seems to work, however I was unable to run the unit tests. When clicking on the play icon in Android Studio I only got 0/0 tests passed.

Also, the changes aren't translated yet. I could be able to add German translations in a future pull request.

UOBW committed:
Added support for the Deluxe Edition (fixes #41)

* Added a setting to enable Deluxe Edition
* The wildfire no longer blanks outsiders if the setting is enabled
* The phoenix logic uses the logic of the deluxe edition if the setting is enabled ("Phoenix is immune to the Book of Changes and may not Blank or be Blanked by any other card" instead of "If this card is BLANKED for any reason, its base strength is reduced to 0 but it retains its suits")
* Reordered the settings page to better fit in the new setting and added some horizontal dividers to improve organization
* Updated the disclaimer on the settings page

UOBW committed:
Updated unit tests

benjamin-klamerek commented:
Hello,

First, I really appreciate when someone adds a contribution to this project (which has been on hiatus for many months now).

I've never worked on the deluxe edition, mostly because it was not available in my language (French).
Now it's available (but the price is a little high... whatever).

Sadly, without the game in my hands, I cannot list ALL the differences compared to the base game and whether it contains the full expansion or not.

I looked really quickly at your PR. I'm not a big fan of the way you added a deluxeEdition flag everywhere (it's a bit aggressive). I admit that it's not an easy task to handle both editions in one app.

Each new card/rule should have dedicated unit tests. I don't know why it's not working on your side. I remember that UI tests require configuration, but simple ones about scoring should run.

I can't promise anything, but I will try to look at it in the next few weeks.

benjamin-klamerek commented:
Also, please note that I won't be able to publish a new app version without increasing the SDK.
I'm just be bit worried about compatibilies (both about smartphones and some librairies)

UOBW commented:
> Sadly, without the game in my hands, I cannot list ALL the differences compared to the base game and whether it contains the full expansion or not.

If have the opposite problem: I have the Deluxe Edition but not the original game, so I compared with this spreadsheet: [boardgamegeek.com/filepage/269150/a-list-of-cards](https://boardgamegeek.com/filepage/269150/a-list-of-cards). The only cards that I could find that changed in the Deluxe Edition, apart from some minor rewordings, are Wildfire, Phoenix, Rangers, and World Tree.

World Tree: The World Tree always gives a 70 points bonus in the Deluxe Edition, even when playing without the Cursed Hoard expansion. I didn't include it in my original pull request because I wasn't sure that you wanted to add it, and I forgot to mention it in the description, but if you want to add that to the app, I can include it in the pull request.

Rangers: The Deluxe Edition added "(except War Dirigible)" after the bonus "CLEARS the word Army from all Penalties". Normally this wouldn't change anything, because the penalty of the War Dirigible is "BLANKED unless with at least one Army", so it doesn't matter if that penalty is cleared. However, if you used the Book of Changes to change the suit of the Rangers the War Dirigible would be blanked in the Deluxe Edition. To me that seems like an unintended side effect of a wording change intended to make the bonus clearer, but it's up to you if you want to include that edge case in the app or not.

Also, the Empress is listed in the spreadshet with a base strength of 10, but I'm pretty sure that it has always been 15.

> I looked really quickly at your PR. I'm not a big fan of the way you added a deluxeEdition flag everywhere (it's a bit aggressive). I admit that it's not an easy task to handle both editions in one app.

Yeah, I know. The problem is that Mirage and Shapeshifter need to know whether the setting is enabled for the card selection dialog, so I added the flag to the `Game` class. The only other alternative I see would be to create an additional two copies of Mirage and Shapeshifter for the Deluxe Edition (base game, cursed hoard, deluxe edition base game, deluxe edition with cursed hoard). If you are worried about about the `Game(deluxeEdition = false)` in the unit tests, I could make the parameter default to `false` so that it can be omitted.

> Each new card/rule should have dedicated unit tests. I don't know why it's not working on your side. I remember that UI tests require configuration, but simple ones about scoring should run.

Okay, I'll try to get the tests running and add some new ones.

> I can't promise anything, but I will try to look at it in the next few weeks.

Sounds good. Thanks for the quick reply.

UOBW commented:
> Also, please note that I won't be able to publish a new app version without increasing the SDK. I'm just be bit worried about compatibilies (both about smartphones and some librairies)

I'm pretty new to Android development and I have no idea how this works with libraries, but as far as I understand if you just increase the `targetSdkVersion` while leaving the `minSdkVersion` unchanged, the app should still be able to run on the same devices that it used to.

UOBW committed:
Updated unit tests & fixed a Phoenix bug

UOBW commented:
Okay, the tests are working now and I added 7 more tests to cover the deluxe edition (6 testing the Phoenix, 1 testing the Wildfire).

I also fixed a bug that I found while writing the tests: `Game::applyBlankingRule()` used to just check if the effects included the object `unblankable` instead of testing if the effect actually returned `Effect.UNBLANKABLE`. This works fine in most cases, but the Phoenix is unblankable only in some cases, so it doesn't use the `unblankable` object but a custom `RuleAboutEffect` implementation.

UOBW commented:
Hi, I just wanted to ask about the state of this pull request and the project as a whole. I have some more ideas on how to improve this app, but I wouldn't want to work on them unless there is a chance that they might be accepted and eventually published to the Play Store, though I understand if there are more important things in your life right now than maintaining this app.
Thanks
